### PR TITLE
Fixes #2548 : Makes InOrder able to verify static methods

### LINKED ISSUE #2548: [PR open] Add feature to verify static methods calls in order
**TL;DR** Mockito doesn't allow verifying static methods calls in order. I have implemented this feature here: https://github.com/mockito/mockito/pull/2549. If possible, I'd like to contribute; or discuss other ways to achieve the same.

---

Hey Mockito team.

This discussion relates to: https://groups.google.com/g/mockito/c/wjd9jIj_oBs. Since I'm not sure that the Google group is active, I'm posting it here for now.

This issue is about verifying static methods calls in order. I'm aware of the InOrder class, but we can't pass a MockedStatic instance. I had a closer look, and we can pass to it a Class instance. But the InOrder interface is not powerful enough to verify that a static method has been called:
```
public interface InOrder {
  <T> T verify(T mock);
  <T> T verify(T mock, VerificationMode mode);
  void verifyNoMoreInteractions();
}
```

To verify static calls, using method references and lambdas, we likely need two more methods in the InOrder interface, with signatures similar to MockedStatic#verify:
```
public interface MockedStatic<T> extends ScopedMock {
  void verify(Verification verification)
  void verify(Verification verification, VerificationMode mode);
  ...
}
```

What do you think of the InOrder interface exposing two methods to deal with the MockedStatic objects that accept both the MockedStatic and the Verification instances to use? Like the following:
```
void verify(MockedStatic<?> mockedStatic, MockedStatic.Verification verification);
void verify(MockedStatic<?> mockedStatic, MockedStatic.Verification verification, VerificationMode mode);
```

I made the above changes, and we can verify static methods, together with instance methods too. I have updated the tests as well in my PR.

Let me know if you are aware of other ways to achieve the same. If it is not doable, then please share your thoughts about the above, and if I can contribute in case.

### PR DESCRIPTION
**Pull request by mirkoalicastro**
This PR makes it possible to verify static methods calls in order. Fixes: #2548 

Example of usage:

```
try (MockedStatic<Foo> mocked = mockStatic(Foo.class)) {
  InOrder inOrder = inOrder(Foo.class);
  mocked.when(Foo::firstMethod).thenReturn("first");
  assertEquals("first", Foo.firstMethod());

  inOrder.verify(mocked, Foo::firstMethod);
}
```
You can also pass a specific `VerificationMode` as third argument to the `InOrder#verify` method.

## Checklist

 - [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/main/.github/CONTRIBUTING.md)
 - [x] PR should be motivated, i.e. what does it fix, why, and if relevant how
 - [x] If possible / relevant include an example in the description, that could help all readers
       including project members to get a better picture of the change
 - [x] Avoid other runtime dependencies
 - [x] Meaningful commit history ; intention is important please rebase your commit history so that each
       commit is meaningful and help the people that will explore a change in 2 years
 - [x] The pull request follows coding style
 - [x] Mention `Fixes #<issue number>` in the description _if relevant_
 - [x] At least one commit should mention `Fixes #<issue number>` _if relevant_

### CODE REVIEW

File: subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+            // when
+            StaticContext.firstMethod();
+
+            // then
+            inOrder.verify(mockedStatic, StaticContext::firstMethod, unsupportedMode);
+        });
+    }
+
+    @Test(expected = VerificationInOrderFailure.class)
```
Comment (TimvdLippe) [inline subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java:93]: Nit: do not use `@Test(expected)` and instead use `org.assertj.core.api.Assertions.assertThatThrownBy`. We recently migrated away from that in #2510 (I should probably check if we can enable ErrorProne to error on this).

**Review by TimvdLippe — CHANGES_REQUESTED** [review]
Thanks for the PR! Overall, the implementation looks solid. I only have a couple of nits with regards to the tests.

My only concern is the difference in the class passed into `inOrder` and the class we are actually interacting with. I guess this currently works, as we are passing in a `Class` in the `List<Object>`, but that feels a bit weird to me. Not sure WDYT?

I think to fix that, we would need a similar method to `objectIsMockToBeVerified`, but in this case verify that the context is correct. However, we are only passing in the lambda, which doesn't necessarily match to the `Class` we put in.

Therefore, we maybe have to look into passing the `mockedStatic` in the `inOrder` method instead and omit it as an argument from the `inOrder.verify` call?

File: subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+import org.junit.Test;
+import org.mockito.InOrder;
+import org.mockito.MockedStatic;
+import org.mockito.exceptions.base.MockitoException;
+import org.mockito.exceptions.misusing.NotAMockException;
+import org.mockito.exceptions.verification.VerificationInOrderFailure;
+import org.mockito.verification.VerificationMode;
+
+public class InOrderVerificationTest {
```
[the remark this answers is not in the document]
Comment (mirkoalicastro) [reply subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java:25]: Definitely, it looks a great test to add. Done!

File: subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            // then
+            if (actual != expected) {
+                fail("Unable to mock static method");
+            }
+            inOrder.verify(mockedStatic, StaticContext::firstMethod);
+            inOrder.verifyNoMoreInteractions();
+        });
+    }
+
+    private void mockedStaticTest(Consumer<MockedStatic<StaticContext>> body) {
```
[the remark this answers is not in the document]
Comment (mirkoalicastro) [reply subprojects/inline/src/test/java/org/mockitoinline/InOrderVerificationTest.java]: Sure, I see your point. I have inlined that in the tests.

**Comment by mirkoalicastro** [comment]
> Thanks for the PR! Overall, the implementation looks solid. I only have a couple of nits with regards to the tests.
> 
> My only concern is the difference in the class passed into `inOrder` and the class we are actually interacting with. I guess this currently works, as we are passing in a `Class` in the `List<Object>`, but that feels a bit weird to me. Not sure WDYT?
> 
> I think to fix that, we would need a similar method to `objectIsMockToBeVerified`, but in this case verify that the context is correct. However, we are only passing in the lambda, which doesn't necessarily match to the `Class` we put in.
> 
> Therefore, we maybe have to look into passing the `mockedStatic` in the `inOrder` method instead and omit it as an argument from the `inOrder.verify` call?

Thanks for your feedback @TimvdLippe ! I have addressed the comments and pushed the changes related to the tests.

I see your concern, and actually, passing the `mockedStatic` to the `inOrder` method was my initial idea. Here is the reason why I moved away from that.
Let's assume we have two static mocks (two different classes), and we want to verify them in order. Then, we'd pass two `MockedStatic` instances to the `inOrder` method.
When it comes to the verification, we'd have `InOrder#verify(Verification)`, where `Verification` is a lambda (no parameters, void return type) because `MockedStatic#verify` deals with a lambda. Given a lambda, we can't assert if it contains a static method reference, or contains just some statements. That is, we can't retrieve the `Class` instance. Without this, we can't select and use the appropriate `MockedStatic` instance to verify the lambda.
Moreover, the above assumes we have the `Class` instance of the `MockedStatic` somehow available in `InOrderImpl`, but currently, that is not the case. Of course, we could add a method to `MockedStatic` that accepts a `Class` and returns true if the `Class` is the same as what is mocking (package private maybe?). 
I see your concern, but I think we have the same situation in the `MockedStatic#verify` methods, where we could pass things not relevant to the context. But I can't see a different way to solve it otherwise, if possible.
WDYT?

**Review by TimvdLippe — APPROVED** [review]
Awesome, thanks for the explanation (makes sense to me) and the nice PR + tests! Merging 🎉

**Pull request merged**